sync/src/main/java/tech/pegasys/teku/sync/AttestationManager.java

```diff
@@ -41,47 +41,48 @@
 
   private final PendingPool<DelayableAttestation> pendingAttestations;
   private final FutureItems<DelayableAttestation> futureAttestations;
+  private final AggregatingAttestationPool aggregatingAttestationPool;
 
   AttestationManager(
-      final EventBus eventBus,
-      final ForkChoiceAttestationProcessor attestationProcessor,
-      final PendingPool<DelayableAttestation> pendingAttestations,
-      final FutureItems<DelayableAttestation> futureAttestations) {
+          final EventBus eventBus,
+          final ForkChoiceAttestationProcessor attestationProcessor,
+          final PendingPool<DelayableAttestation> pendingAttestations,
+          final FutureItems<DelayableAttestation> futureAttestations,
+          final AggregatingAttestationPool aggregatingAttestationPool) {
     this.eventBus = eventBus;
     this.attestationProcessor = attestationProcessor;
     this.pendingAttestations = pendingAttestations;
     this.futureAttestations = futureAttestations;
+    this.aggregatingAttestationPool = aggregatingAttestationPool;
   }
 
   public static AttestationManager create(
       final EventBus eventBus,
       final PendingPool<DelayableAttestation> pendingAttestations,
       final FutureItems<DelayableAttestation> futureAttestations,
-      final ForkChoiceAttestationProcessor forkChoiceAttestationProcessor) {
+      final ForkChoiceAttestationProcessor forkChoiceAttestationProcessor,
+      final AggregatingAttestationPool aggregatingAttestationPool) {
     return new AttestationManager(
-        eventBus, forkChoiceAttestationProcessor, pendingAttestations, futureAttestations);
+        eventBus, forkChoiceAttestationProcessor, pendingAttestations, futureAttestations, aggregatingAttestationPool);
   }
 
   @Subscribe
   @SuppressWarnings("unused")
   private void onGossipedAttestation(final Attestation attestation) {
-    processAttestation(
-        new DelayableAttestation(
-            attestation, () -> eventBus.post(new ProcessedAttestationEvent(attestation))));
+    processAttestation(new DelayableAttestation(attestation, aggregatingAttestationPool::add));
   }
 
   @Subscribe
   @SuppressWarnings("unused")
   private void onAggregateAndProof(final SignedAggregateAndProof aggregateAndProof) {
-    final Attestation aggregate = aggregateAndProof.getMessage().getAggregate();
-    processAttestation(
-        new DelayableAttestation(
-            aggregate, () -> eventBus.post(new ProcessedAggregateEvent(aggregate))));
+    final Attestation aggregateAttestation = aggregateAndProof.getMessage().getAggregate();
+    processAttestation(new DelayableAttestation(aggregateAttestation, aggregatingAttestationPool::add));
+
   }
 
   @Override
   public void onSlot(final UnsignedLong slot) {
-    futureAttestations.prune(slot).forEach(this::processAttestation);
+    futureAttestations.prune(slot).forEach(attestationProcessor::applyAttestationToForkChoice);
```

Comment:
This is where I'd check the `IndexedAttestation` exists and log an error if it doesn't indicating that somehow an unvalidated attestation got into the future pool (then don't process it obviously).

At the moment you could actually make `futureAttestations` just hold `IndexedAttestation` but that's only because we aren't currently calling `delayableAttestation.onAttestationProcessedSuccessfully` for attestations we process from `futureAttestations`.

Reply:
This is exactly what I did after reading your previous comment and having not seen this :D

Reply:
But now we're never even able to call `onAttestationProcessedSuccessfully` when the delayed attestation is actually processed.   I guess the event it fires doesn't actually do anything for now, but we'll probably have to change this back when we want to start gossiping attestations after we process them (if we haven't already).

Although, whether we gossiped something or not seems like a nice thing to track in `ValidatableAttestation` so maybe it will work out...

Reply:
Looking at how all the code has worked out this bit now makes more sense - we're not firing the unused events any more - the on processed event just adds it to the aggregate pool which we've already done for saved attestations.